File: swift/Sources/TelenowSDK/Telenow.swift
// TelenowSDK — iOS/macOS voice client, with auto-reconnect.
//
// Control plane (session init, WebSocket, transcripts, jitter, reconnect) is
// cross-platform and compiles on macOS for CI. Audio I/O uses AVAudioEngine; the
// iOS-only AVAudioSession (voiceChat → hardware AEC) is guarded.

import Foundation
#if canImport(AVFoundation)
import AVFoundation
#endif

public enum CallState: String { case idle, connecting, live, reconnecting, ended, error }

public struct TelenowCallOptions {
    public var token: String?
    public var publicSlug: String?
    public var baseURL: String
    public var variables: [String: String]?
    public var uplinkEncoding: String
    /// Pre-initialized session (your backend called init-web-call with an org
    /// API key). When both are set the SDK skips session init — no token or
    /// publicSlug needed on the device.
    public var sessionId: String?
    public var websocketUrl: String?
    public init(
        token: String? = nil,
        publicSlug: String? = nil,
        baseURL: String = "https://api.telenow.ai",
        variables: [String: String]? = nil,
        uplinkEncoding: String = "mulaw",
        sessionId: String? = nil,
        websocketUrl: String? = nil
    ) {
        self.token = token
        self.publicSlug = publicSlug
        self.baseURL = baseURL
        self.variables = variables
        self.uplinkEncoding = uplinkEncoding
        self.sessionId = sessionId
        self.websocketUrl = websocketUrl
    }
}

public enum TelenowError: Error { case missingCredentials, sessionInit(String), badURL }

struct SessionInfo: Decodable { let sessionId: String; let websocketUrl: String }
private struct Envelope: Decodable { let success: Bool?; let data: SessionInfo?; let error: String? }

/// Exponential backoff with jitter (port of @telenow/client's Reconnector).
struct Reconnector {
    var attempt = 0
    var maxAttempts = 6
    var baseMs = 500.0
    var maxMs = 10000.0
    var jitter = 0.3
    mutating func reset() { attempt = 0 }
    mutating func next(rand: Double) -> Double? {
        if attempt >= maxAttempts { return nil }
        let exp = min(maxMs, baseMs * pow(2.0, Double(attempt)))
        let factor = 1 - jitter + rand * 2 * jitter
        attempt += 1
        return exp * factor
    }
}

public final class TelenowCall {
    public var onState: ((CallState) -> Void)?
    public var onTranscript: ((_ role: String, _ text: String) -> Void)?

    private let options: TelenowCallOptions
    private let urlSession = URLSession(configuration: .default)
    private var ws: URLSessionWebSocketTask?
    private var session: SessionInfo?
    private var recon = Reconnector()
    private var stopped = false
    private var liveAnnounced = false
    private let jitter = AdaptiveJitterBuffer()
    private var clock = 0.0

    #if os(iOS)
    private let engine = AVAudioEngine()
    private let player = AVAudioPlayerNode()
    #endif

    public init(_ options: TelenowCallOptions) { self.options = options }

    public func start() async throws {
        stopped = false
        recon.reset()
        onState?(.connecting)
        let info = try await initSession()
        session = info
        #if os(iOS)
        try startAudio() // started once; survives reconnects
        #endif
        openSocket(info)
    }

    public func stop() {
        stopped = true
        ws?.cancel(with: .normalClosure, reason: nil)
        ws = nil
        #if os(iOS)
        engine.stop()
        #endif
        onState?(.ended)
    }

    public func send(text: String) {
        sendJSON(["event": "text", "text": text, "chat": true])
    }

    // MARK: - Session init
    private func initSession() async throws -> SessionInfo {
        if let sid = options.sessionId, let wsUrl = options.websocketUrl {
            return SessionInfo(sessionId: sid, websocketUrl: wsUrl)
        }
        let base = options.baseURL.hasSuffix("/") ? String(options.baseURL.dropLast()) : options.baseURL
        var request: URLRequest
        let body = try JSONSerialization.data(withJSONObject: ["variables": options.variables ?? [:]])
        if let slug = options.publicSlug {
            guard let url = URL(string: "\(base)/api/public/widget/\(slug)/session") else { throw TelenowError.badURL }
            request = URLRequest(url: url)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        } else if let token = options.token {
            guard let url = URL(string: "\(base)/api/sessions/init-web-call") else { throw TelenowError.badURL }
            request = URLRequest(url: url)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        } else {
            throw TelenowError.missingCredentials
        }
        request.httpMethod = "POST"
        request.httpBody = body
        let (data, _) = try await urlSession.data(for: request)
        let env = try JSONDecoder().decode(Envelope.self, from: data)
        guard env.success == true, let info = env.data else { throw TelenowError.sessionInit(env.error ?? "failed") }
        return info
    }

    // MARK: - WebSocket + reconnect
    private func openSocket(_ info: SessionInfo) {
        guard let url = URL(string: info.websocketUrl) else { onState?(.error); return }
        liveAnnounced = false
        let task = urlSession.webSocketTask(with: url)
        ws = task
        task.resume()
        sendJSON(["event": "start", "sessionId": info.sessionId])
        receive()
    }

    private func receive() {
        ws?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let msg):
                if !self.liveAnnounced {
                    self.liveAnnounced = true
                    self.recon.reset()
                    self.onState?(.live)
                }
                if case .string(let text) = msg,
                   let data = text.data(using: .utf8),
                   let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    self.handle(obj)
                }
                self.receive()
            case .failure:
                self.handleDrop()
            }
        }
    }

    private func handleDrop() {
        if stopped { return }
        guard let info = session, let delayMs = recon.next(rand: Double.random(in: 0...1)) else {
            onState?(.ended)
            return
        }
        onState?(.reconnecting)
        DispatchQueue.global().asyncAfter(deadline: .now() + delayMs / 1000.0) { [weak self] in
            guard let self, !self.stopped else { return }
            self.openSocket(info)
        }
    }

    private func sendJSON(_ obj: [String: Any]) {
        guard let d = try? JSONSerialization.data(withJSONObject: obj),
              let s = String(data: d, encoding: .utf8) else { return }
        ws?.send(.string(s)) { _ in }
    }

    private func handle(_ m: [String: Any]) {
        switch m["event"] as? String {
        case "media":
            guard let b64 = m["data"] as? String, let bytes = Data(base64Encoded: b64) else { return }
            let fmt = (m["format"] as? String) ?? "mulaw"
            let rate = (m["sampleRate"] as? Int) ?? 8000
            let pcm = fmt == "mulaw" ? PCM.mulawToPCM16(bytes) : PCM.le16(bytes)
            if pcm.isEmpty { return }
            let dur = Double(pcm.count) / Double(rate)
            let decision = jitter.schedule(now: clock, frameDur: dur, arrival: clock)
            clock = max(clock, decision.startAt)
            #if os(iOS)
            enqueue(pcm: pcm, rate: rate)
            #endif
        case "clear": // barge-in: drop queued agent audio immediately
            clock = 0
            jitter.reset()
            #if os(iOS)
            player.stop()
            player.play()
            #endif
        case "ping": // echo for server-measured RTT (latency breakdown)
            sendJSON(["event": "pong", "t": m["t"] ?? 0])
        case "transcript":
            onTranscript?((m["role"] as? String) ?? "", (m["text"] as? String) ?? "")
        case "session_end":
            stop()
        default:
            break
        }
    }

    // MARK: - Audio (iOS)
    #if os(iOS)
    public func setMuted(_ muted: Bool) { engine.inputNode.volume = muted ? 0 : 1 }

    private func startAudio() throws {
        let session = AVAudioSession.sharedInstance()
        try session.setCategory(.playAndRecord, mode: .voiceChat, options: [.defaultToSpeaker, .allowBluetooth])
        try session.setActive(true)

        engine.attach(player)
        engine.connect(player, to: engine.mainMixerNode, format: nil)

        let input = engine.inputNode
        let inFormat = input.outputFormat(forBus: 0)
        let target = options.uplinkEncoding == "pcm16" ? 16000.0 : 8000.0
        input.installTap(onBus: 0, bufferSize: 1024, format: inFormat) { [weak self] buffer, _ in
            guard let self, let chan = buffer.floatChannelData?[0] else { return }
            let n = Int(buffer.frameLength)
            var floats = [Float](repeating: 0, count: n)
            for i in 0..<n { floats[i] = chan[i] }
            self.pushUplink(floats, srcRate: inFormat.sampleRate, dstRate: target)
        }
        try engine.start()
        player.play()
    }

    private func pushUplink(_ floats: [Float], srcRate: Double, dstRate: Double) {
        var pcm = [Int16]()
        pcm.reserveCapacity(floats.count)
        for f in floats {
            let c = max(-1, min(1, f))
            pcm.append(Int16((c < 0 ? c * 32768 : c * 32767).rounded()))
        }
        let down = PCM.resampleInt16(pcm, Int(srcRate), Int(dstRate))
        let data: Data = options.uplinkEncoding == "pcm16"
            ? PCM.toLE16(down)
            : Data(down.map { PCM.linear16ToMulawByte($0) })
        sendJSON(["event": "media", "data": data.base64EncodedString()])
    }

    private func enqueue(pcm: [Int16], rate: Int) {
        guard let format = AVAudioFormat(commonFormat: .pcmFormatFloat32, sampleRate: Double(rate), channels: 1, interleaved: false),
              let buf = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: AVAudioFrameCount(pcm.count)) else { return }
        buf.frameLength = AVAudioFrameCount(pcm.count)
        if let ch = buf.floatChannelData?[0] {
            for i in 0..<pcm.count { ch[i] = Float(pcm[i]) / 32768.0 }
        }
        player.scheduleBuffer(buf, completionHandler: nil)
    }
    #else
    public func setMuted(_ muted: Bool) { _ = muted }
    #endif
}

File: android/src/main/java/ai/telenow/sdk/Telenow.kt
// ai.telenow.sdk — Android voice client, with auto-reconnect.
//
// AudioRecord (VOICE_COMMUNICATION → hardware AEC/NS/AGC) + AudioTrack do audio
// I/O; native Kotlin DSP (Dsp.kt) handles codecs + the jitter buffer. The WS
// reconnects with backoff on network drops; audio I/O keeps running across them.
// Requires the Android SDK to compile. Add RECORD_AUDIO + request at runtime.
package ai.telenow.sdk

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.AudioTrack
import android.media.MediaRecorder
import android.util.Base64
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import kotlin.concurrent.thread

enum class CallState { IDLE, CONNECTING, LIVE, RECONNECTING, ENDED, ERROR }

data class TelenowCallOptions(
    val token: String? = null,
    val publicSlug: String? = null,
    val baseUrl: String = "https://api.telenow.ai",
    val variables: Map<String, String>? = null,
    /** "mulaw" (8 kHz, works with the current server) or "pcm16" (16 kHz HD). */
    val uplinkEncoding: String = "mulaw",
    /**
     * Pre-initialized session (from your backend calling init-web-call with an
     * org API key). When both are set, the SDK skips session init entirely —
     * no token/publicSlug needed on the device.
     */
    val sessionId: String? = null,
    val websocketUrl: String? = null,
)

private const val PLAYBACK_RATE = 24000

private class Reconnector(
    val maxAttempts: Int = 6,
    val baseMs: Double = 500.0,
    val maxMs: Double = 10000.0,
    val jitter: Double = 0.3,
) {
    var attempt = 0
    fun reset() { attempt = 0 }
    fun next(rand: Double): Double? {
        if (attempt >= maxAttempts) return null
        val exp = minOf(maxMs, baseMs * Math.pow(2.0, attempt.toDouble()))
        val factor = 1 - jitter + rand * 2 * jitter
        attempt++
        return exp * factor
    }
}

class TelenowCall(private val options: TelenowCallOptions) {
    var onState: ((CallState) -> Unit)? = null
    var onTranscript: ((role: String, text: String) -> Unit)? = null

    private val http = OkHttpClient()
    private var ws: WebSocket? = null
    private val jitter = AdaptiveJitterBuffer()
    private val recon = Reconnector()
    private var clock = 0.0
    @Volatile private var recording = false
    @Volatile private var muted = false
    @Volatile private var stopped = false
    @Volatile private var liveAnnounced = false
    private var sessionId = ""
    private var wsUrl = ""
    private var record: AudioRecord? = null
    private var track: AudioTrack? = null

    private val uplinkRate = if (options.uplinkEncoding == "pcm16") 16000 else 8000

    fun start() {
        stopped = false
        recon.reset()
        onState?.invoke(CallState.CONNECTING)
        thread {
            try {
                val info = initSession()
                sessionId = info.first
                wsUrl = info.second
                startPlayback()
                startCapture() // audio runs once; survives reconnects
                openSocket()
            } catch (e: Exception) {
                onState?.invoke(CallState.ERROR)
            }
        }
    }

    fun setMuted(m: Boolean) { muted = m }

    fun stop() {
        stopped = true
        recording = false
        try { ws?.close(1000, null) } catch (_: Exception) {}
        record?.let { it.stop(); it.release() }; record = null
        track?.let { it.stop(); it.release() }; track = null
        onState?.invoke(CallState.ENDED)
    }

    fun sendText(text: String) {
        ws?.send(JSONObject(mapOf("event" to "text", "text" to text, "chat" to true)).toString())
    }

    private fun initSession(): Pair<String, String> {
        if (options.sessionId != null && options.websocketUrl != null) {
            return options.sessionId to options.websocketUrl
        }
        val base = options.baseUrl.trimEnd('/')
        val urlStr: String
        val headers: Map<String, String>
        when {
            options.publicSlug != null -> {
                urlStr = "$base/api/public/widget/${options.publicSlug}/session"; headers = emptyMap()
            }
            options.token != null -> {
                urlStr = "$base/api/sessions/init-web-call"; headers = mapOf("Authorization" to "Bearer ${options.token}")
            }
            else -> throw IllegalStateException("provide token or publicSlug")
        }
        val conn = URL(urlStr).openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.setRequestProperty("Content-Type", "application/json")
        headers.forEach { (k, v) -> conn.setRequestProperty(k, v) }
        conn.doOutput = true
        val body = JSONObject(mapOf("variables" to (options.variables ?: emptyMap<String, String>()))).toString()
        OutputStreamWriter(conn.outputStream).use { it.write(body) }
        val resp = conn.inputStream.bufferedReader().use { it.readText() }
        val data = JSONObject(resp).getJSONObject("data")
        return data.getString("sessionId") to data.getString("websocketUrl")
    }

    private fun openSocket() {
        liveAnnounced = false
        val req = Request.Builder().url(wsUrl).build()
        ws = http.newWebSocket(req, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                webSocket.send(JSONObject(mapOf("event" to "start", "sessionId" to sessionId)).toString())
            }
            override fun onMessage(webSocket: WebSocket, text: String) {
                if (!liveAnnounced) {
                    liveAnnounced = true
                    recon.reset()
                    onState?.invoke(CallState.LIVE)
                }
                handle(JSONObject(text))
            }
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) = handleDrop()
            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) = handleDrop()
        })
    }

    private fun handleDrop() {
        if (stopped) return
        val delay = recon.next(Math.random())
        if (delay == null) {
            onState?.invoke(CallState.ENDED)
            return
        }
        onState?.invoke(CallState.RECONNECTING)
        thread {
            Thread.sleep(delay.toLong())
            if (!stopped) openSocket()
        }
    }

    private fun handle(m: JSONObject) {
        when (m.optString("event")) {
            "media" -> {
                val b64 = m.optString("data")
                if (b64.isEmpty()) return
                val bytes = Base64.decode(b64, Base64.DEFAULT)
                val fmt = m.optString("format", "mulaw")
                val rate = m.optInt("sampleRate", 8000)
                val pcm = if (fmt == "mulaw") {
                    ShortArray(bytes.size) { Dsp.mulawByteToPcm16(bytes[it].toInt() and 0xff) }
                } else {
                    Dsp.le16ToShorts(bytes)
                }
                if (pcm.isEmpty()) return
                jitter.schedule(clock, pcm.size.toDouble() / rate, clock).also { clock = maxOf(clock, it.startAt) }
                val play = if (rate != PLAYBACK_RATE) Dsp.resample(pcm, rate, PLAYBACK_RATE) else pcm
                track?.write(play, 0, play.size)
            }
            "clear" -> { // barge-in: drop queued agent audio immediately
                track?.let { it.pause(); it.flush(); it.play() }
                jitter.reset()
                clock = 0.0
            }
            "ping" -> ws?.send(JSONObject(mapOf("event" to "pong", "t" to m.opt("t"))).toString())
            "transcript" -> onTranscript?.invoke(m.optString("role"), m.optString("text"))
            "session_end" -> stop()
        }
    }

    private fun startPlayback() {
        val minBuf = AudioTrack.getMinBufferSize(PLAYBACK_RATE, AudioFormat.CHANNEL_OUT_MONO, AudioFormat.ENCODING_PCM_16BIT)
        track = AudioTrack.Builder()
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_VOICE_COMMUNICATION)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                    .build(),
            )
            .setAudioFormat(
                AudioFormat.Builder()
                    .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                    .setSampleRate(PLAYBACK_RATE)
                    .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                    .build(),
            )
            .setBufferSizeInBytes(minBuf * 2)
            .setTransferMode(AudioTrack.MODE_STREAM)
            .build()
        track?.play()
    }

    private fun startCapture() {
        val frame = uplinkRate * 20 / 1000
        val minBuf = AudioRecord.getMinBufferSize(uplinkRate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT)
        record = AudioRecord(
            MediaRecorder.AudioSource.VOICE_COMMUNICATION,
            uplinkRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT,
            maxOf(minBuf, frame * 2) * 2,
        )
        record?.startRecording()
        recording = true
        thread {
            val buf = ShortArray(frame)
            while (recording) {
                val n = record?.read(buf, 0, frame) ?: 0
                if (n <= 0 || muted) continue
                val chunk = if (n == frame) buf else buf.copyOf(n)
                val data = if (options.uplinkEncoding == "pcm16") {
                    Dsp.shortsToLe16(chunk)
                } else {
                    ByteArray(chunk.size) { Dsp.linear16ToMulawByte(chunk[it]) }
                }
                // The server only accepts JSON media envelopes; a bare base64
                // text frame is silently dropped by web_stream's parser.
                ws?.send(
                    JSONObject(
                        mapOf("event" to "media", "data" to Base64.encodeToString(data, Base64.NO_WRAP)),
                    ).toString(),
                )
            }
        }
    }
}
